lib/pbench/server/api/resources/__init__.py

```diff
@@ -397,53 +417,54 @@ def convert_dataset(value: str, _) -> Dataset:
 
 
 def convert_json(value: JSONOBJECT, parameter: "Parameter") -> JSONOBJECT:
-    """
-    Validate a parameter of JSON type. If the Parameter object defines a list
+    """Validate a parameter of JSON type. If the Parameter object defines a list
     of keywords, the JSON key values are validated against that list. If the
-    Parameter key_path attribute is set, then only the first element of a
-    dotted path (e.g., "user" for "user.contact.email") is validated.
+    Parameter key_path attribute is set, then only the first element of a dotted
+    path (e.g., "user" for "user.contact.email") is validated.
 
     Args:
-        value: JSON dict
-        parameter: Supplies list of allowed JSON keys
+        value : JSON dict
+        parameter : Supplies list of allowed JSON keys
 
     Raises:
-        ConversionError: input can't be validated or normalized
+        ConversionError : input can't be validated or normalized
+        KeywordError : unexpected keyword encountered
 
     Returns:
         The JSON dict
     """
     try:
-        if json.loads(json.dumps(value)) == value:
-            if parameter.keywords:
-                bad = []
-                for k in value.keys():
-                    if parameter.key_path:
-                        if not Metadata.is_key_path(k, parameter.keywords):
-                            bad.append(k)
-                    elif k not in parameter.keywords:
-                        bad.append(k)
-                if bad:
-                    raise KeywordError(
-                        parameter, f"JSON key{'s' if len(bad) > 1 else ''}", bad
-                    )
-            return value
-    except JSONDecodeError:
-        pass
-    raise ConversionError(value, "JSON")
+        washed = json.loads(json.dumps(value))
+    except JSONDecodeError as e:
+        raise ConversionError(value, "JSON") from e
+
+    if washed != value:
+        raise ConversionError(value, "JSON")
```

Comment:
Wow: so raise a `ValueError` just to immediately catch it and chain it to a `ConversionError`? While I can see the logic, that's weird and it's going to catch every reader's eye as a likely coding error until they work through it.

Reply:
Yeah, well, the interface to this function declares that it raises `ConversionError`, but that doesn't quite cover the unserializable-JSON case (for which we have a separate, special check); so, while we could expand the interface def to include another exception (e.g., `ValueError`), it seems reasonable/better to just use the single error.  However, raising it without any further context (i.e., with no chained `JSONDecodeError`) seems insufficient.

Now, instead of the `try`/`except`/`raise` path, we could just do a direct `raise` with the `from` specifying the `ValueError`, but I felt like that was even more arcane than what I suggested.

Reply:
Thanks for the suggestion, but I find that too compact.

Reply:
> I find that too compact.

["Too many notes"](https://www.youtube.com/watch?v=dCud8H7z7vU&t=45s), eh?  😆